crates/observe/src/tracing_reload_handler.rs

```diff
@@ -0,0 +1,109 @@
+use {
+    tokio::{
+        io::{AsyncBufReadExt, AsyncWriteExt, BufReader},
+        net::{UnixListener, UnixStream},
+    },
+    tracing_subscriber::{reload, EnvFilter, Registry},
+};
+
+/// Spawns a new thread that listens for connections to a UNIX socket
+/// at "/tmp/log_filter_override_<process_name>_<pid>".
+/// Whenever a line gets writtedn to that socket the reload handler
+/// uses it as the new log filter.
+/// To reset to the original log filter send the message "reset".
+pub(crate) fn spawn_reload_handler(
+    initial_filter: String,
+    reload_handle: reload::Handle<EnvFilter, Registry>,
+) {
+    tokio::spawn(async move {
+        let id = std::process::id();
+        let name = binary_name().unwrap_or_default();
+
+        let socket_path = format!("/tmp/log_filter_override_{name}_{id}.sock");
+        tracing::warn!(file = socket_path, "open log filter reload socket");
+        let handle = SocketHandle {
+            listener: UnixListener::bind(&socket_path).expect("socket handle is unique"),
+            socket_path,
+        };
+
+        loop {
+            handle_connection(&handle.listener, &initial_filter, &reload_handle).await;
+        }
+    });
+}
+
+struct SocketHandle {
+    socket_path: String,
+    listener: UnixListener,
+}
+
+impl Drop for SocketHandle {
+    fn drop(&mut self) {
+        let _ = std::fs::remove_file(&self.socket_path);
+    }
+}
+
+fn binary_name() -> Option<String> {
+    Some(
+        std::env::current_exe()
+            .ok()?
+            .file_name()?
+            .to_str()?
+            .to_string(),
+    )
+}
+
+async fn handle_connection(
+    listener: &UnixListener,
+    initial_filter: &str,
+    reload_handle: &reload::Handle<EnvFilter, Registry>,
+) {
+    let Ok((mut socket, _addr)) = listener.accept().await else {
+        tracing::warn!("failed to accept UNIX socket connection");
+        return;
+    };
+
+    let _ = socket
+        .write_all(format!("log filter on process startup was: {initial_filter:?}\n",).as_bytes())
+        .await;
+
+    loop {
```

Comment:
how do we break from this loop? IIUC, once we connect to it to change the logs, we cannot connect again to change it back.

Reply:
Good catch. I believe I test that at some point but it broke while refactoring the code. 👍

Reply:
Apparently the `BufReader` returns an empty string when the stream gets terminated. Used that to break the loop.
Tested it locally.